File: CS/mrtbPainter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors.Drawing;
using DevExpress.Skins;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.Utils.Drawing;
using System.Drawing;

namespace CustomRangeTrackBarControl
{
    class MultiRangePainter : RangeTrackBarPainter
    {
        public MultiRangePainter()
        {
        }
    }

    class MultipleTrackBarObjectPainter : SkinRangeTrackBarObjectPainter
    {
        public MultipleTrackBarObjectPainter(DevExpress.Skins.ISkinProvider provider)
            : base(provider)
        { }

        protected virtual void DrawThumbs(TrackBarObjectInfoArgs e)
        {
            MultiRangeViewInfo vi = (e.ViewInfo as MultiRangeViewInfo);
            Point p = vi.MaxThumbPos;
            RangeList list = (vi.EditValue as RangeList);
            for (int i = 0; i < list.Count; i++ )
            {
                TrackBarObjectInfoArgs args = new TrackBarObjectInfoArgs(vi.TrackInfo.Appearance, vi);
                args.Cache = e.ViewInfo.TrackInfo.Cache;
                vi.CurrentIndex++;
                DrawRangeHighlight(args, e.ViewInfo.TrackBarHelper.Rotate(e.ViewInfo.TrackLineRect));
                DrawThumb(args);
            }
            vi.CurrentIndex = 0;
        }
        
        public override void DrawMaxThumb(TrackBarObjectInfoArgs e, bool bMirror)
        {
            MultiRangeViewInfo ri = e.ViewInfo as MultiRangeViewInfo;
            SkinElementInfo maxInfo = GetMaxThumbInfo(ri, e.State);
            if (maxInfo == null) return;
            maxInfo.Bounds = GetVerticalThumbRectangle(e.ViewInfo, ri.MaxThumbBounds);
            if (ri.CurrentIndex == ri.DraggedThumb && ri.PressedInfo.HitTest == EditHitTest.Button2)
                maxInfo.State = ObjectState.Pressed;
            new RotateObjectPaintHelper().DrawRotated(e.Cache, maxInfo, SkinElementPainter.Default, GetRotateAngle(e.ViewInfo), true);
        }

        public override void DrawMinThumb(TrackBarObjectInfoArgs e, bool bMirror)
        {
            MultiRangeViewInfo ri = e.ViewInfo as MultiRangeViewInfo;
            SkinElementInfo minInfo = GetMinThumbInfo(ri, e.State);
            if (minInfo == null) return;
            if (ri.CurrentIndex == ri.DraggedThumb && ri.PressedInfo.HitTest == EditHitTest.Button)
                minInfo.State = ObjectState.Pressed;
            minInfo.Bounds = GetMinVerticalThumbRectangle(ri, ri.MinThumbBounds);
            new RotateObjectPaintHelper().DrawRotated(e.Cache, minInfo, SkinElementPainter.Default, GetRotateAngle(e.ViewInfo), true);
        }

        public override void DrawObject(DevExpress.Utils.Drawing.ObjectInfoArgs e)
        {
            TrackBarObjectInfoArgs tbe = e as TrackBarObjectInfoArgs;
            DrawBackground(tbe);
            DrawTrackLine(tbe);
            if (this.AllowTick(tbe.ViewInfo))
                DrawPoints(tbe);
            (tbe.ViewInfo as MultiRangeViewInfo).CurrentIndex = -1;
            DrawThumbs(tbe);
            if (tbe.ViewInfo.Item.ShowLabels) DrawLabels(tbe);
        }

        protected virtual void DrawRangeHighlight(TrackBarObjectInfoArgs e, Rectangle bounds)
        {
            SkinElementInfo info = new SkinElementInfo(GetTrack(e.ViewInfo), bounds);
            Rectangle filled = GetFilledRect(e);
            GraphicsClipState clipState = e.Cache.ClipInfo.SaveAndSetClip(filled);
            if (e.ViewInfo.Item.HighlightSelectedRange)
                info.ImageIndex += 1;
            new RotateObjectPaintHelper().DrawRotated(e.Cache, info, SkinElementPainter.Default, GetTrackLineRotateAngle(e.ViewInfo));
            e.Cache.ClipInfo.RestoreClipRelease(clipState);
        }

        protected override void DrawSkinTrackLineCore(TrackBarObjectInfoArgs e, Rectangle bounds)
        {
            SkinElementInfo info = new SkinElementInfo(GetTrack(e.ViewInfo), bounds);
            info.ImageIndex = e.State == ObjectState.Disabled ? 2 : 0;
            new RotateObjectPaintHelper().DrawRotated(e.Cache, info, SkinElementPainter.Default, GetTrackLineRotateAngle(e.ViewInfo));
        }

        public override Rectangle GetFilledRect(TrackBarObjectInfoArgs e)
        {
            MultiRangeViewInfo ri = e.ViewInfo as MultiRangeViewInfo;
            Rectangle rect = ri.TrackLineRect;
            rect.X = ri.ThumbsPos[ri.CurrentIndex].Minimum.X;
            rect.Width = ri.ThumbsPos[ri.CurrentIndex].Maximum.X - ri.ThumbsPos[ri.CurrentIndex].Minimum.X;
            return e.ViewInfo.TrackBarHelper.Rotate(rect);
        }

        protected override TrackBarInfoCalculator GetCalculator(TrackBarViewInfo viewInfo)
        {
            return new CustomCalculator(viewInfo as RangeTrackBarViewInfo, this);
        }
    }
}

File: VB/mrtbPainter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.Skins
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.Utils.Drawing
Imports System.Drawing

Namespace CustomRangeTrackBarControl
	Friend Class MultiRangePainter
		Inherits RangeTrackBarPainter
		Public Sub New()
		End Sub
	End Class

	Friend Class MultipleTrackBarObjectPainter
		Inherits SkinRangeTrackBarObjectPainter
		Public Sub New(ByVal provider As DevExpress.Skins.ISkinProvider)
			MyBase.New(provider)
		End Sub

		Protected Overridable Sub DrawThumbs(ByVal e As TrackBarObjectInfoArgs)
			Dim vi As MultiRangeViewInfo = (TryCast(e.ViewInfo, MultiRangeViewInfo))
			Dim p As Point = vi.MaxThumbPos
			Dim list As RangeList = (TryCast(vi.EditValue, RangeList))
			For i As Integer = 0 To list.Count - 1
				Dim args As New TrackBarObjectInfoArgs(vi.TrackInfo.Appearance, vi)
				args.Cache = e.ViewInfo.TrackInfo.Cache
				vi.CurrentIndex += 1
				DrawRangeHighlight(args, e.ViewInfo.TrackBarHelper.Rotate(e.ViewInfo.TrackLineRect))
				DrawThumb(args)
			Next i
			vi.CurrentIndex = 0
		End Sub

		Public Overrides Sub DrawMaxThumb(ByVal e As TrackBarObjectInfoArgs, ByVal bMirror As Boolean)
			Dim ri As MultiRangeViewInfo = TryCast(e.ViewInfo, MultiRangeViewInfo)
			Dim maxInfo As SkinElementInfo = GetMaxThumbInfo(ri, e.State)
			If maxInfo Is Nothing Then
				Return
			End If
			maxInfo.Bounds = GetVerticalThumbRectangle(e.ViewInfo, ri.MaxThumbBounds)
			If ri.CurrentIndex = ri.DraggedThumb AndAlso ri.PressedInfo.HitTest = EditHitTest.Button2 Then
				maxInfo.State = ObjectState.Pressed
			End If
			CType(New RotateObjectPaintHelper(), RotateObjectPaintHelper).DrawRotated(e.Cache, maxInfo, SkinElementPainter.Default, GetRotateAngle(e.ViewInfo), True)
		End Sub

		Public Overrides Sub DrawMinThumb(ByVal e As TrackBarObjectInfoArgs, ByVal bMirror As Boolean)
			Dim ri As MultiRangeViewInfo = TryCast(e.ViewInfo, MultiRangeViewInfo)
			Dim minInfo As SkinElementInfo = GetMinThumbInfo(ri, e.State)
			If minInfo Is Nothing Then
				Return
			End If
			If ri.CurrentIndex = ri.DraggedThumb AndAlso ri.PressedInfo.HitTest = EditHitTest.Button Then
				minInfo.State = ObjectState.Pressed
			End If
			minInfo.Bounds = GetMinVerticalThumbRectangle(ri, ri.MinThumbBounds)
			CType(New RotateObjectPaintHelper(), RotateObjectPaintHelper).DrawRotated(e.Cache, minInfo, SkinElementPainter.Default, GetRotateAngle(e.ViewInfo), True)
		End Sub

		Public Overrides Overloads Sub DrawObject(ByVal e As DevExpress.Utils.Drawing.ObjectInfoArgs)
			Dim tbe As TrackBarObjectInfoArgs = TryCast(e, TrackBarObjectInfoArgs)
			DrawBackground(tbe)
			DrawTrackLine(tbe)
			If Me.AllowTick(tbe.ViewInfo) Then
				DrawPoints(tbe)
			End If
			TryCast(tbe.ViewInfo, MultiRangeViewInfo).CurrentIndex = -1
			DrawThumbs(tbe)
			If tbe.ViewInfo.Item.ShowLabels Then
				DrawLabels(tbe)
			End If
		End Sub

		Protected Overridable Sub DrawRangeHighlight(ByVal e As TrackBarObjectInfoArgs, ByVal bounds As Rectangle)
			Dim info As New SkinElementInfo(GetTrack(e.ViewInfo), bounds)
			Dim filled As Rectangle = GetFilledRect(e)
			Dim clipState As GraphicsClipState = e.Cache.ClipInfo.SaveAndSetClip(filled)
			If e.ViewInfo.Item.HighlightSelectedRange Then
				info.ImageIndex += 1
			End If
			CType(New RotateObjectPaintHelper(), RotateObjectPaintHelper).DrawRotated(e.Cache, info, SkinElementPainter.Default, GetTrackLineRotateAngle(e.ViewInfo))
			e.Cache.ClipInfo.RestoreClipRelease(clipState)
		End Sub

		Protected Overrides Sub DrawSkinTrackLineCore(ByVal e As TrackBarObjectInfoArgs, ByVal bounds As Rectangle)
			Dim info As New SkinElementInfo(GetTrack(e.ViewInfo), bounds)
			info.ImageIndex = If(e.State = ObjectState.Disabled, 2, 0)
			CType(New RotateObjectPaintHelper(), RotateObjectPaintHelper).DrawRotated(e.Cache, info, SkinElementPainter.Default, GetTrackLineRotateAngle(e.ViewInfo))
		End Sub

		Public Overrides Function GetFilledRect(ByVal e As TrackBarObjectInfoArgs) As Rectangle
			Dim ri As MultiRangeViewInfo = TryCast(e.ViewInfo, MultiRangeViewInfo)
			Dim rect As Rectangle = ri.TrackLineRect
			rect.X = ri.ThumbsPos(ri.CurrentIndex).Minimum.X
			rect.Width = ri.ThumbsPos(ri.CurrentIndex).Maximum.X - ri.ThumbsPos(ri.CurrentIndex).Minimum.X
			Return e.ViewInfo.TrackBarHelper.Rotate(rect)
		End Function

		Protected Overrides Function GetCalculator(ByVal viewInfo As TrackBarViewInfo) As TrackBarInfoCalculator
			Return New CustomCalculator(TryCast(viewInfo, RangeTrackBarViewInfo), Me)
		End Function
	End Class
End Namespace
